### ISSUE TITLE
[Quince testing] <add course mode from django admin>: <Getting 403 error while adding course mode from django admin>

### ISSUE THREAD

**Issue by fayyazahmed66** [openedx/wg-build-test-release#325]
### Release

quince

### Expected behavior

The user should be able to add course mode without any issues.

### Actual behavior

I get 403 when I try to save course mode.

### Steps to reproduce

1. Go to https://quince.demo.edly.io/admin/course_modes/coursemode/add/
2. Select any course and add its mode
3. click on the save button

### Additional information

<img width="1440" alt="Screenshot 2023-11-03 at 4 43 01 PM" src="https://github.com/openedx/wg-build-test-release/assets/55684124/511d1d65-5a36-4dd7-a5ed-1c353b4ffe2a">

**Comment by regisb** [comment]
It should be noted that the "csrftoken" appears to be correctly set (value = `QJvNDZ9GQQFw3aydr6L4MLWpQPXJKzbD`):

![Screenshot from 2023-11-06 09-35-14](https://github.com/openedx/wg-build-test-release/assets/44319/6ce314a2-95c9-452f-ba78-a5ea40504bf1)

The "csrfmiddlewaretoken" request parameter is also included in the POST request, but it has a different value from the cookie (`sdJux230jFcULjJES9oPtrYtDZSNwFPw8M470R2wZlHgEj7H95ZJ52KIjEFm64QZ`):

![image](https://github.com/openedx/wg-build-test-release/assets/44319/bac181c7-5f1d-4d83-a278-9bb2697986f6)


Maybe this discrepancy is causing the 403 response? 

In any case, the solution should NOT be to ignore CSRF parameters for the LMS host. CSRF checking is an important security feature and it should not be bypassed.

**Comment by iamsobanjaved** [comment]
Not familiar with the way the tutor is set up but this may help.
In Django 4.0, the Origin header is now checked and Django recommends adding the origin to the `CSRF_TRUSTED_ORIGINS` if the request is coming from a subdomain, more details here: https://docs.djangoproject.com/en/4.2/releases/4.0/#csrf-trusted-origins-changes

**Comment by regisb** [comment]
@iamsobanjaved the request is not coming from a subdomain. The request is coming from the same host as the LMS.

**Comment by iamsobanjaved** [comment]
Can we set `DEBUG = True` in this test instance temporarily to get the reason behind the error?

**Comment by regisb** [comment]
Sure. There you go:

![image](https://github.com/openedx/wg-build-test-release/assets/44319/b422b346-2056-4bbe-a0a2-5028c47c82da)

**Comment by mariajgrimaldi** [comment]
label: quince testing

**Comment by regisb** [comment]
For the record, I am unable to reproduce the issue in development mode. (`tutor dev start`)

**Comment by mariajgrimaldi** [comment]
I tried a different POST from the LMS and it's also failing:
![image](https://github.com/openedx/wg-build-test-release/assets/64440265/c406e022-624c-42ab-888d-ed71df4d4028)

But it's not failing from studio: 
![image](https://github.com/openedx/wg-build-test-release/assets/64440265/0b66d10e-0c71-4717-a865-b99dcce4cef3)

**Comment by regisb** [comment]
I'll write here my investigations.

In `django.middleware.csrf._origin_verified` the test `if request_origin == good_origin` is failing because `request.secure()` is false, which is surprising. What is even more surprising is that `request.scheme == "https"`. Yet, the implementation of `is_secure` is unambiguous: https://github.com/django/django/blob/main/django/http/request.py#L268

Turns out that the `request.is_secure` method is MANUALLY OVERRIDDEN by edx-platform (all caps to emphasise my growing surprise). Right here: https://github.com/openedx/edx-platform/blob/e51c01bf4ec69d254e4fd4fb79a341f99309fd38/openedx/core/djangoapps/cors_csrf/helpers.py#L75 (`skip_cross_domain_referer_check` context manager)

The call to `skip_cross_domain_referer_check` that concerns us is the one from the `CorsCSRFMiddleware` middleware: https://github.com/openedx/edx-platform/blob/89c8f21b94d420a3b58e33e3890bf4410a787487/openedx/core/djangoapps/cors_csrf/middleware.py#L76

What is strange to me is that this middleware has not changed for many years... And it's been enabled in the LMS for a long time. So maybe this is a false lead?

I need to log out. If someone wants to pick up from there, be my guest :hugs:

**Comment by timmc-edx** [comment]
Yeah, that cors_csrf middleware is kind of scary. I filed a DEPR for it a couple of days ago (https://discuss.openedx.org/t/deprecation-removal-cors-csrf-middleware-and-utilities-edx-platform-33627/11577) after I saw that code during review of <https://github.com/openedx/edx-platform/pull/33226>.

**Comment by regisb** [comment]
There are two candidate solutions. It's important to realize that both are temporary, so let's not close this issue.

1. Add LMS_HOST to `CSRF_TRUSTED_ORIGINS` (or `CSRF_TRUSTED_ORIGINS_WITH_SCHEME`): I think this might resolve the issue but it might open a new attack vector.
2. Remove the `'openedx.core.djangoapps.cors_csrf.middleware.CorsCSRFMiddleware'` entry from the `MIDDLEWARE` setting in the lms.envs.production module. I have no idea if there are unforeseen consequences with this change. This might break the marketing sites, and probably other sites as well.

I implemented solution 2 so that testing can resume. But we must realise that the issue is caused by [the upgrade to Django 4.2 in the Quince branch](https://github.com/openedx/edx-platform/pull/33516). This backport DOES NOT EXIST in master. So we are in a strange situation where Quince runs Django 4.2, but master runs Django 3. This is highly unusual.

**Comment by mariajgrimaldi** [comment]
In Django 4.x, a new check for the CSRF middleware was introduced: 
https://github.com/django/django/pull/13829
https://github.com/django/django/blob/main/docs/releases/4.0.txt#L233-L238
That might explain why it's not failing in Django 3.x

**Comment by mariajgrimaldi** [comment]
@regisb: it looks like in edx.org they already added the host to `CSRF_TRUSTED_ORIGINS` so they only added the missing scheme https://openedx.atlassian.net/wiki/spaces/COMM/pages/3726802953/Pending+Release+Quince#Operational-%E2%9A%99%EF%B8%8F

**Comment by regisb** [comment]
> maaaybe it wouldn't fail even if they were using django 4.0

I think you're right. But I'm not sure that disabling CSRF for the LMS is the right choice.

**Comment by regisb** [comment]
:clap: :clap: :clap: @zawan-ila :clap: :clap: :clap: 

Nailed it! Following your suggestion, I added to the LMS production settings:

```
MIDDLEWARE.remove('django.middleware.csrf.CsrfViewMiddleware')
MIDDLEWARE.insert(MIDDLEWARE.index('corsheaders.middleware.CorsMiddleware'), 'django.middleware.csrf.CsrfViewMiddleware')
```

Lo and behold, this resolves the issue.

Thus, the right fix would be to move the Django middleware up in the LMS common settings. @zawan-ila would you like to open a PR or should I do it? Once the fix is in master we should also backport it to Quince.

Thank you so much for researching this @zawan-ila, we were truly out of our depth here.

**Comment by zawan-ila** [comment]
@regisb Feel free to open any pull requests as you see appropriate.

**Comment by regisb** [comment]
Who wants to review this? https://github.com/openedx/edx-platform/pull/33727

**Comment by mariajgrimaldi** [comment]
@regisb: I'll do it.

**Comment by mariajgrimaldi** [comment]
This was fixed by the PR linked above.

**Comment by fayyazahmed66** [comment]
@mariajgrimaldi  Verified. Thanks.
